Pipfile scripts with $PIPENV_PROJECT_DIR fail to work

Be sure to check the existing issues (both open and closed!), and make sure you are running the latest version of Pipenv.

Check the [diagnose documentation](https://pipenv.pypa.io/en/latest/diagnose.html) for common issues before posting! We may close your issue if it is very similar to one of them. Please be considerate, or be on your way.

Make sure to mention your debugging experience if the documented solution failed.


### Issue description

Pipfile does not actually resolve variable usages in scripts. Which, while this limitation is exactly as described in [this docs snippet](https://github.com/pypa/pipenv/blob/fd90cda55fd78a7301097f65d09d094356764102/docs/scripts.md?plain=1#L90-L104), is implied to work in a limited scope based on [this other docs snippet](https://github.com/pypa/pipenv/blob/fd90cda55fd78a7301097f65d09d094356764102/docs/configuration.md?plain=1#L101-L119).

The fact that there is no ability to use $PIPENV_PROJECT_DIR means that currently, the feature currently fails to meet the adaquately meet 2nd use case detailed to justify its addition (https://github.com/pypa/pipenv/pull/6497).

### Expected result

```
username@device:~/Documents/Python/min_repro$ pipenv run mypy
src/test.py:3: error: No return value expected  [return-value]
Found 1 error in 1 file (checked 1 source file)
```

### Actual result

```
username@device:~/Documents/Python/min_repro$ pipenv run mypy --verbose
usage: mypy [-h] [-v] [-V] [more options; see below]
                        [-m MODULE] [-p PACKAGE] [-c PROGRAM_TEXT] [files ...]
mypy: error: Cannot find config file '$PIPENV_PROJECT_DIR/mypy.ini'
```

### Steps to replicate

Provide the steps to replicate (which usually at least includes the commands and the Pipfile).
File Structure:

```
min_repro/
├── mypy.ini
├── Pipfile
├── Pipfile.lock
└── src/
    └── test.py
```

<details>
  <summary>mypy.ini</summary>

```ini
[mypy]
python_version = 3.14
```
</details>

<details>
  <summary>Pipfile</summary>

```toml
[[source]]
url = "https://pypi.org/simple"
verify_ssl = true
name = "pypi"

[packages]
mypy = "*"

[requires]
python_version = "3.14"
[scripts]
mypy = "mypy --config-file=$PIPENV_PROJECT_DIR/mypy.ini $PIPENV_PROJECT_DIR/src"
```
</details>

<details>
  <summary>src/test.py</summary>

```python3
def foo() -> None:
    """An intentionally wrong annotation for mypy to find."""
    return "bar"
```
</details>

run command:

`$ pipenv run mypy`

-------------------------------------------------------------------------------

<details><summary>$ pipenv --support</summary>

Pipenv version: `'2026.5.2'`

Pipenv location: `'/home/username/.virtualenvs/pipenv-venv/lib/python3.14/site-packages/pipenv'`

Python location: `'/home/username/.virtualenvs/pipenv-venv/bin/python3.14'`

OS Name: `'posix'`

User pip version: `'26.0.1'`

user Python installations found:

  - `3.14.3`: `/home/username/.virtualenvs/pipenv-venv/bin/python3`
  - `3.14.3`: `/home/username/.virtualenvs/pipenv-venv/bin/python`
  - `3.14.3`: `/home/username/.pyenv/shims/python`
  - `3.12.3`: `/usr/bin/python3`
  - `3.12.3`: `/bin/python3`
  - `3.7.17`: `/usr/bin/python3.7m`
  - `3.7.17`: `/usr/bin/python3.7`
  - `3.7.17`: `/bin/python3.7m`
  - `3.7.17`: `/bin/python3.7`
  - `2.5.4`: `/home/username/.pyenv/versions/2.5.4/bin/python`
  - `2.5.4`: `/home/username/.pyenv/versions/2.5.4/bin/python2.5`

PEP 508 Information:

```
{'implementation_name': 'cpython',
 'implementation_version': '3.14.3',
 'os_name': 'posix',
 'platform_machine': 'x86_64',
 'platform_python_implementation': 'CPython',
 'platform_release': '6.17.0-20-generic',
 'platform_system': 'Linux',
 'platform_version': '#20~24.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Mar 19 '
                     '01:28:37 UTC 2',
 'python_full_version': '3.14.3',
 'python_version': '3.14',
 'sys_platform': 'linux'}
```

System environment variables:

  - `SHELL`
  - `SESSION_MANAGER`
  - `QT_ACCESSIBILITY`
  - `COLORTERM`
  - `PYENV_SHELL`
  - `XDG_CONFIG_DIRS`
  - `NVM_INC`
  - `XDG_MENU_PREFIX`
  - `GNOME_DESKTOP_SESSION_ID`
  - `GNOME_SHELL_SESSION_MODE`
  - `DOTNET_ROOT`
  - `SSH_AUTH_SOCK`
  - `MEMORY_PRESSURE_WRITE`
  - `XMODIFIERS`
  - `DESKTOP_SESSION`
  - `GTK_MODULES`
  - `PWD`
  - `LOGNAME`
  - `XDG_SESSION_DESKTOP`
  - `XDG_SESSION_TYPE`
  - `GPG_AGENT_INFO`
  - `SYSTEMD_EXEC_PID`
  - `XAUTHORITY`
  - `WINDOWPATH`
  - `VIRTUALENVWRAPPER_SCRIPT`
  - `HOME`
  - `USERNAME`
  - `LANG`
  - `LS_COLORS`
  - `_VIRTUALENVWRAPPER_API`
  - `XDG_CURRENT_DESKTOP`
  - `MEMORY_PRESSURE_WATCH`
  - `VTE_VERSION`
  - `GNOME_TERMINAL_SCREEN`
  - `QTWEBENGINE_DICTIONARIES_PATH`
  - `PIPENV_IGNORE_VIRTUALENVS`
  - `DOTNET_BUNDLE_EXTRACT_BASE_DIR`
  - `NVM_DIR`
  - `SGX_AESM_ADDR`
  - `LESSCLOSE`
  - `XDG_SESSION_CLASS`
  - `TERM`
  - `LESSOPEN`
  - `LIBVIRT_DEFAULT_URI`
  - `USER`
  - `GNOME_TERMINAL_SERVICE`
  - `CLUTTER_DISABLE_MIPMAPPED_TEXT`
  - `DISPLAY`
  - `SHLVL`
  - `NVM_CD_FLAGS`
  - `GSM_SKIP_SSH_AGENT_WORKAROUND`
  - `QT_IM_MODULE`
  - `XDG_RUNTIME_DIR`
  - `PYENV_ROOT`
  - `DEBUGINFOD_URLS`
  - `XDG_DATA_DIRS`
  - `PATH`
  - `GDMSESSION`
  - `DBUS_SESSION_BUS_ADDRESS`
  - `NVM_BIN`
  - `OLDPWD`
  - `_`
  - `PIP_DISABLE_PIP_VERSION_CHECK`
  - `PYTHONDONTWRITEBYTECODE`

Pipenv–specific environment variables:

 - `PIPENV_IGNORE_VIRTUALENVS`: `1`

Debug–specific environment variables:

  - `PATH`: `/home/username/.virtualenvs/pipenv-venv/bin:/home/username/.nvm/versions/node/v22.16.0/bin:/home/username/.pyenv/shims:/home/username/.pyenv/bin:/home/username/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin:/home/username/.dotnet/tools:/home/username/.modular/bin:/opt/mssql-tools18/bin:/usr/share/dotnet:/usr/share/dotnet/tools:/usr/local/go/bin:/home/username/programs/JetBrains Rider-2025.1.1/bin`
  - `SHELL`: `/bin/bash`
  - `LANG`: `en_US.UTF-8`
  - `PWD`: `/home/username/Documents/Python/min_repro`


---------------------------

Contents of `Pipfile` ('/home/username/Documents/Python/min_repro/Pipfile'):

```toml
[[source]]
url = "https://pypi.org/simple"
verify_ssl = true
name = "pypi"

[packages]
mypy = "*"

[dev-packages]

[requires]
python_version = "3.14"

[scripts]
mypy = "mypy --config-file=$PIPENV_PROJECT_DIR/mypy.ini $PIPENV_PROJECT_DIR/src"

```


Contents of `Pipfile.lock` ('/home/username/Documents/Python/min_repro/Pipfile.lock'):

```json
{
    "_meta": {
        "hash": {
            "sha256": "062d4b1da2b3b138bc3e597f765c1e340a46e13cb0911b1c118f89ce8ea4528e"
        },
        "pipfile-spec": 6,
        "requires": {
            "python_version": "3.14"
        },
        "sources": [
            {
                "name": "pypi",
                "url": "https://pypi.org/simple",
                "verify_ssl": true
            }
        ]
    },
    "default": {
        "librt": {
            "hashes": [
                "sha256:07cf11f769831186eeac424376e6189f20ace4f7263e2134bdb9757340d84d4d",
                "sha256:0a09c2f5869649101738653a9b7ab70cf045a1105ac66cbb8f4055e61df78f2d",
                "sha256:0a1be03168b2691ba61927e299b352a6315189199ca18a57b733f86cb3cc8d38",
                "sha256:0b73e4266307e51c95e09c0750b7ec383c561d2e97d58e473f6f6a209952fbb8",
                "sha256:15cb151e52a044f06e54ac7f7b47adbfc89b5c8e2b63e1175a9d587c43e8942a",
                "sha256:194fc1a32e1e21fe809d38b5faea66cc65eaa00217c8901fbdb99866938adbdb",
                "sha256:1bf465d1e5b0a27713862441f6467b5ab76385f4ecf8f1f3a44f8aa3c695b4b6",
                "sha256:1c587494461ebd42229d0f1739f3aa34237dd9980623ecf1be8d3bcba79f4499",
                "sha256:224b9727eb8bc188bc3bcf29d969dba0cd61b01d9bac80c41575520cc4baabb2",
                "sha256:22a904cbdb678f7cb348c90d543d3c52f581663d687992fee47fd566dcbf5285",
                "sha256:2b8f5d00b49818f4e2b1667db994488b045835e0ac16fe2f924f3871bd2b8ac5",
                "sha256:2c3786f0f4490a5cd87f1ed6cefae833ad6b1060d52044ce0434a2e85893afd0",
                "sha256:2d03fa4fd277a7974c1978c92c374c57f44edeee163d147b477b143446ad1bf6",
                "sha256:2f8e12706dcb8ff6b3ed57514a19e45c49ad00bcd423e87b2b2e4b5f64578443",
                "sha256:3be322a15ee5e70b93b7a59cfd074614f22cc8c9ff18bd27f474e79137ea8d3b",
                "sha256:3cc2917258e131ae5f958a4d872e07555b51cb7466a43433218061c74ef33745",
                "sha256:3f05d145df35dca5056a8bc3838e940efebd893a54b3e19b2dda39ceaa299bcb",
                "sha256:3fe56e80badb66fdcde06bef81bbaa5bfcf6fbd7aefb86222d9e369c38c6b228",
                "sha256:42ff8a962554c350d4a83cf47d9b7b78b0e6ff7943e87df7cdfc97c07f3c016f",
                "sha256:451daa98463b7695b0a30aa56bf637831ea559e7b8101ac2ef6382e8eb15e29c",
                "sha256:465814ab157986acb9dfa5ccd7df944be5eefc0d08d31ec6e8d88bc71251d845",
                "sha256:4c4d0440a3a8e31d962340c3e1cc3fc9ee7febd34c8d8f770d06adb947779ea5",
                "sha256:4e3dda8345307fd7306db0ed0cb109a63a2c85ba780eb9dc2d09b2049a931f9c",
                "sha256:5093043afb226ecfa1400120d1ebd4442b4f99977783e4f4f7248879009b227f",
                "sha256:5112c2fb7c2eefefaeaf5c97fec81343ef44ee86a30dcfaa8223822fba6467b4",
                "sha256:527b5b820b47a09e09829051452bb0d1dd2122261254e2a6f674d12f1d793d54",
                "sha256:54d412e47c21b85865676ed0724e37a89e9593c2eee1e7367adf85bfad56ffb1",
                "sha256:56d65b583cf43b8cf4c8fbe1e1da20fa3076cc32a1149a141507af1062718236",
                "sha256:5ca8e133d799c948db2ab1afc081c333a825b5540475164726dcbf73537e5c2f",
                "sha256:603138ee838ee1583f1b960b62d5d0007845c5c423feb68e44648b1359014e27",
                "sha256:63c12efcd160e1d14da11af0c46c0217473e1e0d2ae1acbccc83f561ea4c2a7b",
                "sha256:657f8ba7b9eaaa82759a104137aed2a3ef7bc46ccfd43e0d89b04005b3e0a4cc",
                "sha256:66b58fed90a545328e80d575467244de3741e088c1af928f0b489ebec3ef3858",
                "sha256:6788207daa0c19955d2b668f3294a368d19f67d9b5f274553fd073c1260cbb9f",
                "sha256:703f4ae36d6240bfe24f542bac784c7e4194ec49c3ba5a994d02891649e2d85b",
                "sha256:7202bdcac47d3a708271c4304a474a8605a4a9a4a709e954bf2d3241140aa938",
                "sha256:756775d25ec8345b837ab52effee3ad2f3b2dfd6bbee3e3f029c517bd5d8f05a",
                "sha256:78042f6facfd98ecb25e9829c7e37cce23363d9d7c83bc5f72702c5059eb082b",
                "sha256:789fff71757facc0738e8d89e3b84e4f0251c1c975e85e81b152cdaca927cc2d",
                "sha256:7bc30ad339f4e1a01d4917d645e522a0bc0030644d8973f6346397c93ba1503f",
                "sha256:7d429bdd4ac0ab17c8e4a8af0ed2a7440b16eba474909ab357131018fe8c7e71",
                "sha256:7d5c8a5929ac325729f6119802070b561f4db793dffc45e9ac750992a4ed4d22",
                "sha256:7e6274fd33fc5b2a14d41c9119629d3ff395849d8bcbc80cf637d9e8d2034da8",
                "sha256:80b25c7b570a86c03b5da69e665809deb39265476e8e21d96a9328f9762f9990",
                "sha256:81107843ed1836874b46b310f9b1816abcb89912af627868522461c3b7333c0f",
                "sha256:8494cfc61e03542f2d381e71804990b3931175a29b9278fdb4a5459948778dc2",
                "sha256:850d6d03177e52700af605fd60db7f37dcb89782049a149674d1a9649c2138fd",
                "sha256:8c6bc1384d9738781cfd41d09ad7f6e8af13cfea2c75ece6bd6d2566cdea2076",
                "sha256:90904fac73c478f4b83f4ed96c99c8208b75e6f9a8a1910548f69a00f1eaa671",
                "sha256:90e6d5420fc8a300518d4d2288154ff45005e920425c22cbbfe8330f3f754bd9",
                "sha256:928bd06eca2c2bbf4349e5b817f837509b0604342e65a502de1d50a7570afd15",
                "sha256:9b3e3bc363f71bda1639a4ee593cb78f7fbfeacc73411ec0d4c92f00730010a4",
                "sha256:9edcc35d1cae9fd5320171b1a838c7da8a5c968af31e82ecc3dff30b4be0957f",
                "sha256:9fcb461fbf70654a52a7cc670e606f04449e2374c199b1825f754e16dacfedd8",
                "sha256:a0951822531e7aee6e0dfb556b30d5ee36bbe234faf60c20a16c01be3530869d",
                "sha256:a361c9434a64d70a7dbb771d1de302c0cc9f13c0bffe1cf7e642152814b35265",
                "sha256:a4b25c6c25cac5d0d9d6d6da855195b254e0021e513e0249f0e3b444dc6e0e61",
                "sha256:a5af136bfba820d592f86c67affcef9b3ff4d4360ac3255e341e964489b48519",
                "sha256:a81eea9b999b985e4bacc650c4312805ea7008fd5e45e1bf221310176a7bcb3a",
                "sha256:a9c63e04d003bc0fb6a03b348018b9a3002f98268200e22cc80f146beac5dc40",
                "sha256:aa95738a68cedd3a6f5492feddc513e2e166b50602958139e47bbdd82da0f5a7",
                "sha256:b0a2040f801406b93657a70b72fa12311063a319fee72ce98e1524da7200171f",
                "sha256:b8bd70d5d816566a580d193326912f4a76ec2d28a97dc4cd4cc831c0af8e330e",
                "sha256:b8da9f8035bb417770b1e1610526d87ad4fc58a2804dc4d79c53f6d2cf5a6eb9",
                "sha256:bc5518873822d2faa8ebdd2c1a4d7c8ef47b01a058495ab7924cb65bdbf5fc9a",
                "sha256:c0d620e74897f8c2613b3c4e2e9c1e422eb46d2ddd07df540784d44117836af3",
                "sha256:c2640e23d2b7c98796f123ffd95cf2022c7777aa8a4a3b98b36c570d37e85eee",
                "sha256:c81aef782380f0f13ead670aae01825eb653b44b046aa0e5ebbb79f76ed4aa11",
                "sha256:d4d16b608a1c43d7e33142099a75cd93af482dadce0bf82421e91cad077157f4",
                "sha256:d69fc39e627908f4c03297d5a88d9284b73f4d90b424461e32e8c2485e21c283",
                "sha256:d9da80e5b04acce03ced8ba6479a71c2a2edf535c2acc0d09c80d2f80f3bad15",
                "sha256:dd2c7e082b0b92e1baa4da28163a808672485617bc855cc22a2fd06978fa9084",
                "sha256:de7dac64e3eb832ffc7b840eb8f52f76420cde1b845be51b2a0f6b870890645e",
                "sha256:e0785c2fb4a81e1aece366aa3e2e039f4a4d7d21aaaded5227d7f3c703427882",
                "sha256:e306d956cfa027fe041585f02a1602c32bfa6bb8ebea4899d373383295a6c62f",
                "sha256:e78fb7419e07d98c2af4b8567b72b3eaf8cb05caad642e9963465569c8b2d87e",
                "sha256:e9002e98dcb1c0a66723592520decd86238ddcef168b37ff6cfb559200b4b774",
                "sha256:e94cbc6ad9a6aeea46d775cbb11f361022f778a9cc8cc90af653d3a594b057ce",
                "sha256:eea1b54943475f51698f85fa230c65ccac769f1e603b981be060ac5763d90927",
                "sha256:f100bfe2acf8a3689af9d0cc660d89f17286c9c795f9f18f7b62dd1a6b247ae6",
                "sha256:f162af66a2ed3f7d1d161a82ca584efd15acd9c1cff190a373458c32f7d42118",
                "sha256:f24b90b0e0c8cc9491fb1693ae91fe17cb7963153a1946395acdbdd5818429a4",
                "sha256:f29b68cd9714531672db62cc54f6e8ff981900f824d13fa0e00749189e13778e",
                "sha256:f38bc489037eca88d6ebefc9c4d41a4e07c8e8b4de5188a9e6d290273ad7ebb1",
                "sha256:f3fd278f5e6bf7c75ccd6d12344eb686cc020712683363b66f46ac79d37c799f",
                "sha256:f4003f70c56a5addd6aa0897f200dd59afd3bf7bcd5b3cce46dd21f925743bc2",
                "sha256:f48c963a76d71b9d7927eb817b543d0dccd52ab6648b99d37bd54f4cd475d856",
                "sha256:f819e0c6413e259a17a7c0d49f97f405abadd3c2a316a3b46c6440b7dbbedbb1",
                "sha256:fc5758e2b7a56532dc33e3c544d78cbaa9ecf0a0f2a2da2df882c1d6b99a317f",
                "sha256:fcbdf2a9ca24e87bbebb47f1fe34e531ef06f104f98c9ccfc953a3f3344c567a"
            ],
            "markers": "python_version >= '3.9'",
            "version": "==0.9.0"
        },
        "mypy": {
            "hashes": [
                "sha256:0c64e5973df366b747646fc98da921f9d6eba9716d57d1db94a83c026a08e0fb",
                "sha256:0deb80d062b2479f2c87ae568f89845afc71d11bc41b04179e58165fd9f31e98",
                "sha256:1e1c12f6d2db3d78b909b5f77513c11eb7f2dd2782b96a3ab6dffc7d44575c99",
                "sha256:20175a1c0f49863946ec20b7f63255768058ac4f07d2b9ded6a6b46cfb5a9100",
                "sha256:29752dbbf8cc53f89f6ac096d363314333045c257c9c75cbd189ca2de0455744",
                "sha256:2a4102b03bb7481d9a91a6da8d174740c9c8c4401024684b9ca3b7cc5e49852f",
                "sha256:2de3dcea53babc1c3237a19002bc3d228ce1833278f093b8d619e06e7cc79609",
                "sha256:34397cdced6b90b836e38182076049fdb41424322e0b0728c946b0939ebdf9f6",
                "sha256:4077797a273e56e8843d001e9dfe4ba10e33323d6ade647ff260e5cd97d9758c",
                "sha256:419413398fe250aae057fd2fe50166b61077083c9b82754c341cf4fd73038f30",
                "sha256:4b6481b228d072315b053210b01ac320e1be243dc17f9e5887ef167f23f5fae4",
                "sha256:4dbfcf869f6b0517f70cf0030ba6ea1d6645e132337a7d5204a18d8d5636c02b",
                "sha256:4fef51b01e638974a6e69885687e9bd40c8d1e09a6cd291cca0619625cf1f558",
                "sha256:52b176444e2e5054dfcbcb8c75b0b719865c96247b37407184bbfca5c353f2c2",
                "sha256:56908d7e08318d39f85b1f0c6cfd47b0cac1a130da677630dac0de3e0623e102",
                "sha256:5a65aa591af023864fd08a97da9974e919452cfe19cb146c8a5dc692626445dc",
                "sha256:688c3312e5dadb573a2c69c82af3a298d43ecf9e6d264e0f95df960b5f6ac19c",
                "sha256:6e2b469efd811707bc530fd1effef0f5d6eebcb7fe376affae69025da4b979a2",
                "sha256:7488448de6007cd5177c6cea0517ac33b4c0f5ee9b5e9f2be51ce75511a85517",
                "sha256:785b08db19c9f214dc37d65f7c165d19a30fcecb48abfa30f31b01b5acaabb58",
                "sha256:7b0e817b518bff7facd7f85ea05b643ad8bdcce684cf29784987b0a7c8e1f997",
                "sha256:803203d2b6ea644982c644895c2f78b28d0e208bba7b27d9b921e0ec5eb207c6",
                "sha256:89dce27e142d25ffbc154c1819383b69f2e9234dc4ed4766f42e0e8cb264ab5c",
                "sha256:913485a03f1bcf5d279409a9d2b9ed565c151f61c09f29991e5faa14033da4c8",
                "sha256:97d7b9a485b40f8ca425460e89bf1da2814625b2da627c0dcc6aa46c92631d14",
                "sha256:9bcb8aa397ff0093c824182fd76a935a9ba7ad097fcbef80ae89bf6c1731d8ec",
                "sha256:9d56a78b646f2e3daa865bc70cd5ec5a46c50045801ca8ff17a0c43abc97e3ee",
                "sha256:a5da6976f20cae27059ea8d0c86e7cef3de720e04c4bb9ee18e3690fdb792066",
                "sha256:a94c5a76ab46c5e6257c7972b6c8cff0574201ca7dc05647e33e795d78680563",
                "sha256:a95a9248b0c6fd933a442c03c3b113c3b61320086b88e2c444676d3fd1ca3330",
                "sha256:baf593f2765fa3a6b1ef95807dbaa3d25b594f6a52adcc506a6b9cb115e1be67",
                "sha256:bb9c2fa06887e21d6a3a868762acb82aec34e2c6fd0174064f27c93ede68ad15",
                "sha256:bba9ad231e92a3e424b3e56b65aa17704993425bba97e302c832f9466bb85bac",
                "sha256:c3bae4f855d965b5453784300c12ffc63a548304ac7f99e55d4dc7c898673aa3",
                "sha256:c566c3a88b6ece59b3d70f65bedef17304f48eb52ff040a6a18214e1917b3254",
                "sha256:cdecf62abcc4292500d7858aeae87a1f8f1150f4c4dd08fb0b336ee79b2a6df3",
                "sha256:cf5a4db6dca263010e2c7bff081c89383c72d187ba2cf4c44759aac970e2f0c4",
                "sha256:d52ad8d78522da1d308789df651ee5379088e77c76cb1994858d40a426b343b9",
                "sha256:e061b58443f1736f8a37c48978d7ab581636d6ab03e3d4f99e3fa90463bb9382",
                "sha256:e2877a02380adfcdbc69071a0f74d6e9dbbf593c0dc9d174e1f223ffd5281943",
                "sha256:e73c07f23009962885c197ccb9b41356a30cc0e5a1d0c2ea8fd8fb1362d7f924",
                "sha256:e8222c26daaafd9e8626dec58ae36029f82585890589576f769a650dd20fd665",
                "sha256:edfbfca868cdd6bd8d974a60f8a3682f5565d3f5c99b327640cedd24c4264026",
                "sha256:f376e37f9bf2a946872fc5fd1199c99310748e3c26c7a26683f13f8bdb756cbd"
            ],
            "index": "pypi",
            "markers": "python_version >= '3.10'",
            "version": "==1.20.2"
        },
        "mypy-extensions": {
            "hashes": [
                "sha256:1be4cccdb0f2482337c4743e60421de3a356cd97508abadd57d47403e94f5505",
                "sha256:52e68efc3284861e772bbcd66823fde5ae21fd2fdb51c62a211403730b916558"
            ],
            "markers": "python_version >= '3.8'",
            "version": "==1.1.0"
        },
        "pathspec": {
            "hashes": [
                "sha256:574b128f7456bd899045ccd142dd446af7e6cfd0072d63ad73fbc55fbb4aaa42",
                "sha256:f5d7c555da02fd8dde3e4a2354b6aba817a89112fa8f333f7917a2a4834dd080"
            ],
            "markers": "python_version >= '3.9'",
            "version": "==1.1.0"
        },
        "typing-extensions": {
            "hashes": [
                "sha256:0cea48d173cc12fa28ecabc3b837ea3cf6f38c6d1136f85cbaaf598984861466",
                "sha256:f0fa19c6845758ab08074a0cfa8b7aecb71c999ca73d62883bc25cc018c4e548"
            ],
            "markers": "python_version >= '3.9'",
            "version": "==4.15.0"
        }
    },
    "develop": {}
}

```
</details>

--- a/news/6652.bugfix.rst
+++ b/news/6652.bugfix.rst
@@ -1,0 +1,3 @@
+``pipenv run`` now expands ``$PIPENV_PROJECT_DIR`` and other Pipenv-managed
+environment variables inside Pipfile script arguments before direct command
+execution, so project-relative script paths resolve correctly.

--- a/pipenv/routines/shell.py
+++ b/pipenv/routines/shell.py
@@ -1,11 +1,10 @@
 import os
 import subprocess
 import sys
-from os.path import expandvars
 
 from pipenv.utils import err
 from pipenv.utils.project import ensure_project
-from pipenv.utils.shell import cmd_list_to_shell, system_which
+from pipenv.utils.shell import cmd_list_to_shell, safe_expandvars, system_which
 from pipenv.utils.virtualenv import virtualenv_scripts_dir
 
 
@@ -172,7 +171,8 @@ def _run_script_sequence(script, env, verbose=False):
         sub, sub_inline_env = sub.with_extracted_env_vars()
         sub_string_env = {**string_env, **sub_inline_env}
 
-        cmd_args = [sub.command] + [expandvars(arg) for arg in sub.args]
+        expanded_args = [safe_expandvars(arg, env=sub_string_env) for arg in sub.args]
+        cmd_args = [sub.command] + expanded_args
         if verbose:
             err.print(f"$ {cmd_list_to_shell(cmd_args)}", style="cyan")
 
@@ -180,7 +180,7 @@ def _run_script_sequence(script, env, verbose=False):
             # On Windows mirror _launch_windows_subprocess: try direct
             # CreateProcess first, fall back to shell=True.
             command_path = system_which(sub.command, path=path)
-            sub.cmd_args[1:] = [expandvars(arg) for arg in sub.args]
+            sub.cmd_args[1:] = expanded_args
             if command_path:
                 try:
                     result = subprocess.run(
@@ -200,7 +200,7 @@ def _run_script_sequence(script, env, verbose=False):
             command_path = system_which(sub.command, path=path)
             if command_path:
                 result = subprocess.run(
-                    [command_path, *(expandvars(arg) for arg in sub.args)],
+                    [command_path, *expanded_args],
                     env=sub_string_env,
                     check=False,
                 )
@@ -224,19 +224,20 @@ def do_run_posix(project, script, command, env):
 
     # Ensure all environment variables are strings
     string_env = {k: str(v) for k, v in env.items() if v is not None}
+    expanded_args = [safe_expandvars(arg, env=string_env) for arg in script.args]
 
     if command_path:
         # Command found in PATH, use os.execve for direct execution
         os.execve(
             command_path,
-            [command_path, *(expandvars(arg) for arg in script.args)],
+            [command_path, *expanded_args],
             string_env,
         )
     else:
         # Command not found in PATH, maybe it's a shell builtin (cd, echo, export, etc.)
         # Fall back to running through the shell, similar to Windows behavior.
         # See: https://github.com/pypa/pipenv/issues/6186
-        cmd_args = [script.command] + [expandvars(arg) for arg in script.args]
+        cmd_args = [script.command] + expanded_args
         cmd_string = cmd_list_to_shell(cmd_args)
         result = subprocess.run(cmd_string, shell=True, env=string_env, check=False)
         sys.exit(result.returncode)
@@ -255,7 +256,7 @@ def _launch_windows_subprocess(script, env):
     # Ensure all environment variables are strings
     string_env = {k: str(v) for k, v in env.items() if v is not None}
     options = {"universal_newlines": True, "env": string_env}
-    script.cmd_args[1:] = [expandvars(arg) for arg in script.args]
+    script.cmd_args[1:] = [safe_expandvars(arg, env=string_env) for arg in script.args]
 
     # Command not found, maybe this is a shell built-in?
     if not command:

--- a/pipenv/utils/shell.py
+++ b/pipenv/utils/shell.py
@@ -217,20 +217,46 @@ def escape_cmd(cmd):
     return cmd
 
 
+# $VAR or ${VAR} — matches os.path.expandvars (posixpath) semantics.
+_VAR_PROG = re.compile(r"\$(\w+|\{[^}]*\})")
+# %VAR% — Windows-only form, also recognised by os.path.expandvars on nt.
+_VAR_PROG_PERCENT = re.compile(r"%([^%]*)%")
+
+
+def _expand_with_mapping(text, mapping):
+    """Expand ``$VAR`` / ``${VAR}`` (and ``%VAR%`` on Windows) against an
+    explicit mapping, leaving unknown names unchanged. Pure-Python — does not
+    touch ``os.environ`` and is therefore thread-safe."""
+
+    def _sub(match):
+        name = match.group(1)
+        if name.startswith("{") and name.endswith("}"):
+            name = name[1:-1]
+        if name in mapping:
+            return str(mapping[name])
+        return match.group(0)
+
+    result = _VAR_PROG.sub(_sub, text)
+    if os.name == "nt":
+        result = _VAR_PROG_PERCENT.sub(_sub, result)
+    return result
+
+
 def safe_expandvars(value, env=None):
     """
     Expand environment variables in a string value, do nothing for non-strings.
 
-    Note: pathlib.Path doesn't have an expandvars method, so we still use os.path.expandvars
-    for the actual expansion.
+    When ``env`` is provided, expansion uses *only* that mapping — variables
+    inherited from the ambient process environment are not visible. This avoids
+    leaking vars from an outer pipenv invocation into a nested run.
     """
     if env is not None:
-        with temp_environ():
-            # Expand against exactly the provided env mapping so inherited
-            # variables from an outer pipenv invocation do not leak in.
-            os.environ.clear()
-            os.environ.update({k: str(v) for k, v in env.items() if v is not None})
-            return safe_expandvars(value)
+        mapping = {k: v for k, v in env.items() if v is not None}
+        if isinstance(value, str):
+            return _expand_with_mapping(value, mapping)
+        if isinstance(value, Path):
+            return Path(_expand_with_mapping(str(value), mapping))
+        return value
     if isinstance(value, str):
         return os.path.expandvars(value)
     # Handle Path objects

--- a/pipenv/utils/shell.py
+++ b/pipenv/utils/shell.py
@@ -217,13 +217,17 @@ def escape_cmd(cmd):
     return cmd
 
 
-def safe_expandvars(value):
+def safe_expandvars(value, env=None):
     """
     Expand environment variables in a string value, do nothing for non-strings.
 
     Note: pathlib.Path doesn't have an expandvars method, so we still use os.path.expandvars
     for the actual expansion.
     """
+    if env is not None:
+        with temp_environ():
+            os.environ.update({k: str(v) for k, v in env.items() if v is not None})
+            return safe_expandvars(value)
     if isinstance(value, str):
         return os.path.expandvars(value)
     # Handle Path objects

--- a/pipenv/utils/shell.py
+++ b/pipenv/utils/shell.py
@@ -226,6 +226,9 @@ def safe_expandvars(value, env=None):
     """
     if env is not None:
         with temp_environ():
+            # Expand against exactly the provided env mapping so inherited
+            # variables from an outer pipenv invocation do not leak in.
+            os.environ.clear()
             os.environ.update({k: str(v) for k, v in env.items() if v is not None})
             return safe_expandvars(value)
     if isinstance(value, str):

--- a/tests/integration/test_run.py
+++ b/tests/integration/test_run.py
@@ -233,20 +233,20 @@ def test_run_script_expands_pipenv_project_dir(pipenv_instance_pypi):
     with pipenv_instance_pypi() as p:
         p.pipenv("install")
 
-        marker_path = os.path.join(p.path, "marker.txt")
-        with open(marker_path, "w") as f:
+        with open(os.path.join(p.path, "marker.txt"), "w") as f:
             f.write("marker")
 
         with open(p.pipfile_path, "w") as f:
             f.write(
                 "[scripts]\n"
-                'show_project_file = "python -c \\"import pathlib, sys; path = pathlib.Path(sys.argv[1]); print(path); print(path.is_file())\\" $PIPENV_PROJECT_DIR/marker.txt"\n'
+                'show_project_file = "python -c \\"import pathlib, sys; print(pathlib.Path(sys.argv[1]).read_text())\\" $PIPENV_PROJECT_DIR/marker.txt"\n'
             )
 
-        c = p.pipenv("run show_project_file")
+        with temp_environ():
+            os.environ["PIPENV_PROJECT_DIR"] = "/definitely/wrong"
+            c = p.pipenv("run show_project_file")
         assert c.returncode == 0, c.stderr
-        lines = [line.strip() for line in c.stdout.splitlines()]
-        assert lines == [marker_path, "True"]
+        assert c.stdout.strip() == "marker"
 
 
 @pytest.mark.run

--- a/tests/integration/test_run.py
+++ b/tests/integration/test_run.py
@@ -226,33 +226,6 @@ def test_run_inline_env_vars(pipenv_instance_pypi):
         assert c.returncode == 0, c.stderr
         assert "hello world" in c.stdout
 
-
-@pytest.mark.run
-@pytest.mark.skip_windows
-def test_run_script_expands_pipenv_project_dir(pipenv_instance_pypi):
-    with pipenv_instance_pypi() as p:
-        p.pipenv("install")
-
-        with open(os.path.join(p.path, "marker.txt"), "w") as f:
-            f.write("marker")
-
-        with open(p.pipfile_path, "w") as f:
-            f.write(
-                "[scripts]\n"
-                'show_project_file = "python -c \\"import pathlib, sys; print(pathlib.Path(sys.argv[1]).read_text())\\" $PIPENV_PROJECT_DIR/marker.txt"\n'
-            )
-
-        with temp_environ():
-            # CI runs the test suite under ``pipenv run pytest``, which injects
-            # the outer project's ``PIPENV_PROJECT_DIR`` into worker envs.
-            # Clear only that marker so the temp project can define its own
-            # without tripping nested-virtualenv courtesy behavior.
-            os.environ.pop("PIPENV_PROJECT_DIR", None)
-            c = p.pipenv("run show_project_file")
-        assert c.returncode == 0, c.stderr
-        assert c.stdout.strip() == "marker"
-
-
 @pytest.mark.run
 @pytest.mark.skip_windows
 def test_run_shell_builtins(pipenv_instance_pypi):

--- a/tests/integration/test_run.py
+++ b/tests/integration/test_run.py
@@ -229,6 +229,28 @@ def test_run_inline_env_vars(pipenv_instance_pypi):
 
 @pytest.mark.run
 @pytest.mark.skip_windows
+def test_run_script_expands_pipenv_project_dir(pipenv_instance_pypi):
+    with pipenv_instance_pypi() as p:
+        p.pipenv("install")
+
+        marker_path = os.path.join(p.path, "marker.txt")
+        with open(marker_path, "w") as f:
+            f.write("marker")
+
+        with open(p.pipfile_path, "w") as f:
+            f.write(
+                "[scripts]\n"
+                'show_project_file = "python -c \\"import pathlib, sys; path = pathlib.Path(sys.argv[1]); print(path); print(path.is_file())\\" $PIPENV_PROJECT_DIR/marker.txt"\n'
+            )
+
+        c = p.pipenv("run show_project_file")
+        assert c.returncode == 0, c.stderr
+        lines = [line.strip() for line in c.stdout.splitlines()]
+        assert lines == [marker_path, "True"]
+
+
+@pytest.mark.run
+@pytest.mark.skip_windows
 def test_run_shell_builtins(pipenv_instance_pypi):
     """Test that shell builtins work with pipenv run.
 

--- a/tests/integration/test_run.py
+++ b/tests/integration/test_run.py
@@ -244,9 +244,9 @@ def test_run_script_expands_pipenv_project_dir(pipenv_instance_pypi):
 
         with temp_environ():
             # CI runs the test suite under ``pipenv run pytest``, which injects
-            # outer-project markers into the worker environment. Drop them so
-            # this regression exercises the temp project directly.
-            os.environ.pop("PIPENV_ACTIVE", None)
+            # the outer project's ``PIPENV_PROJECT_DIR`` into worker envs.
+            # Clear only that marker so the temp project can define its own
+            # without tripping nested-virtualenv courtesy behavior.
             os.environ.pop("PIPENV_PROJECT_DIR", None)
             c = p.pipenv("run show_project_file")
         assert c.returncode == 0, c.stderr

--- a/tests/integration/test_run.py
+++ b/tests/integration/test_run.py
@@ -243,7 +243,11 @@ def test_run_script_expands_pipenv_project_dir(pipenv_instance_pypi):
             )
 
         with temp_environ():
-            os.environ["PIPENV_PROJECT_DIR"] = "/definitely/wrong"
+            # CI runs the test suite under ``pipenv run pytest``, which injects
+            # outer-project markers into the worker environment. Drop them so
+            # this regression exercises the temp project directly.
+            os.environ.pop("PIPENV_ACTIVE", None)
+            os.environ.pop("PIPENV_PROJECT_DIR", None)
             c = p.pipenv("run show_project_file")
         assert c.returncode == 0, c.stderr
         assert c.stdout.strip() == "marker"

--- a/tests/unit/test_core.py
+++ b/tests/unit/test_core.py
@@ -4,7 +4,9 @@ from unittest.mock import MagicMock, PropertyMock, patch
 
 import pytest
 
+from pipenv.cmdparse import Script
 from pipenv.project import NON_CATEGORY_SECTIONS
+from pipenv.routines.shell import do_run_posix
 from pipenv.shells import _get_activate_script, _get_deactivate_wrapper_script
 from pipenv.utils.environment import load_dot_env
 from pipenv.utils.shell import temp_environ
@@ -144,6 +146,24 @@ def test_load_dot_env_suppresses_message_when_pipenv_active(monkeypatch, capsys,
         assert os.environ[key] == val
         # But the "Loading .env" message should be suppressed
         assert "Loading .env" not in err
+
+
+@pytest.mark.core
+def test_do_run_posix_expands_project_dir_from_child_env():
+    script = Script("python", ["-c", "print('ok')", "$PIPENV_PROJECT_DIR/marker.txt"])
+    env = {"PATH": "/usr/bin", "PIPENV_PROJECT_DIR": "/tmp/project-root"}
+
+    with temp_environ():
+        os.environ["PIPENV_PROJECT_DIR"] = "/outer/project"
+        with patch("pipenv.routines.shell.system_which", return_value="/usr/bin/python"):
+            with patch("os.execve") as execve:
+                do_run_posix(project=MagicMock(), script=script, command="python", env=env)
+
+    execve.assert_called_once_with(
+        "/usr/bin/python",
+        ["/usr/bin/python", "-c", "print('ok')", "/tmp/project-root/marker.txt"],
+        {"PATH": "/usr/bin", "PIPENV_PROJECT_DIR": "/tmp/project-root"},
+    )
 
 
 @pytest.mark.core

--- a/tests/unit/test_utils.py
+++ b/tests/unit/test_utils.py
@@ -2259,10 +2259,12 @@ def test_expand_url_credentials_unset_var_left_unchanged():
 @pytest.mark.utils
 def test_safe_expandvars_with_explicit_env_does_not_leak_ambient_vars(monkeypatch):
     monkeypatch.setenv("PIPENV_PROJECT_DIR", "/outer/project")
+    path_value = "/usr/bin:/bin"
+    monkeypatch.setenv("PATH", path_value)
 
     result = shell.safe_expandvars(
         "$PIPENV_PROJECT_DIR/marker.txt",
-        env={"PATH": os.environ["PATH"]},
+        env={"PATH": path_value},
     )
 
     assert result == "$PIPENV_PROJECT_DIR/marker.txt"

--- a/tests/unit/test_utils.py
+++ b/tests/unit/test_utils.py
@@ -2256,6 +2256,18 @@ def test_expand_url_credentials_unset_var_left_unchanged():
     assert "${NONEXISTENT_VAR_12345}@" in result
 
 
+@pytest.mark.utils
+def test_safe_expandvars_with_explicit_env_does_not_leak_ambient_vars(monkeypatch):
+    monkeypatch.setenv("PIPENV_PROJECT_DIR", "/outer/project")
+
+    result = shell.safe_expandvars(
+        "$PIPENV_PROJECT_DIR/marker.txt",
+        env={"PATH": os.environ["PATH"]},
+    )
+
+    assert result == "$PIPENV_PROJECT_DIR/marker.txt"
+
+
 class TestTargetMarkerEnvironment:
     """Tests for _target_marker_environment (GH-6647).
 

